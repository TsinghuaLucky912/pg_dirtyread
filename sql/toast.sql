create table toast (
    description text,
    data text
);

insert into toast values ('short inline', 'xxx');
insert into toast values ('long inline uncompressed', repeat('x', 200));

alter table toast alter column data set storage external;
insert into toast values ('external uncompressed', repeat('0123456789 8< ', 200));

alter table toast alter column data set storage extended;
insert into toast values ('inline compressed pglz', repeat('0123456789 8< ', 200));
insert into toast values ('extended compressed pglz', repeat('0123456789 8< ', 20000));

alter table toast alter column data set compression lz4;
insert into toast values ('inline compressed lz4', repeat('0123456789 8< ', 200));
insert into toast values ('extended compressed lz4', repeat('0123456789 8< ', 50000));

select description, pg_column_size(data), substring(data, 1, 50) as data from toast;
delete from toast;

-- toasted values are uncompressed after pg_dirtyread
select description, pg_column_size(data), substring(data, 1, 50) as data from pg_dirtyread('toast') as (description text, data text);
