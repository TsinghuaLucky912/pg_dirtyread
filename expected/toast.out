create table toast (
    description text,
    data text
);
insert into toast values ('short inline', 'xxx');
insert into toast values ('long inline uncompressed', repeat('x', 200));
alter table toast alter column data set storage external;
insert into toast values ('external uncompressed', repeat('0123456789 8< ', 200));
alter table toast alter column data set storage extended;
insert into toast values ('inline compressed pglz', repeat('0123456789 8< ', 200));
insert into toast values ('extended compressed pglz', repeat('0123456789 8< ', 20000));
alter table toast alter column data set compression lz4;
insert into toast values ('inline compressed lz4', repeat('0123456789 8< ', 200));
insert into toast values ('extended compressed lz4', repeat('0123456789 8< ', 50000));
select description, pg_column_size(data), substring(data, 1, 50) as data from toast;
       description        | pg_column_size |                        data                        
--------------------------+----------------+----------------------------------------------------
 short inline             |              4 | xxx
 long inline uncompressed |            204 | xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 external uncompressed    |           2800 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
 inline compressed pglz   |             59 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
 extended compressed pglz |           3226 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
 inline compressed lz4    |             42 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
 extended compressed lz4  |           2772 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
(7 rows)

delete from toast;
-- toasted values are uncompressed after pg_dirtyread
select description, pg_column_size(data), substring(data, 1, 50) as data from pg_dirtyread('toast') as (description text, data text);
       description        | pg_column_size |                        data                        
--------------------------+----------------+----------------------------------------------------
 short inline             |              4 | xxx
 long inline uncompressed |            204 | xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 external uncompressed    |           2804 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
 inline compressed pglz   |             59 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
 extended compressed pglz |         280004 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
 inline compressed lz4    |             42 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
 extended compressed lz4  |         700004 | 0123456789 8< 0123456789 8< 0123456789 8< 01234567
(7 rows)

